### PR TITLE
replace np.prod with utilities.shape.size_from_shape to improve performance

### PR DESCRIPTION
**Pull request by michael-123123**
This pull request follows the discussion in issue #1784, which replaces calls to np.prod in expression.py and reshape.py with calls to functools.reduce, as recommended by @phschiele 

## Description
Performance improvement of hstack.

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [x] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/utilities/shape.py [hunk 1]
```diff
@@ -145,3 +145,25 @@ def mul_shapes(lh_shape: Tuple[int, ...], rh_shape: Tuple[int, ...]) -> Tuple[in
...
+    shape : tuple
+        a tuple of integers describing the shape of an object
+
+    Returns
+    -------
+    int
+        The size of an object corresponding to shape.
+    """
+    if shape:
+        return reduce(mul, shape)
```
[the remark this answers is not in the document]
Comment (michael-123123) [reply cvxpy/utilities/shape.py:169]: you're right. need to import both reduce and mul.
I'm sorry - I thought that pycharm took care of it - but somehow it was overlooked.
will fix.

**Comment by keithbriggs** [comment]
A couple more comments:
1. The "else" at line 168 is not needed.
2. The comment "Compute the size of a given shape by multiplying the size of its dimensions" does not make sense to me. Maybe it should be more like "Compute the size of a given shape by multiplying the sizes of each axis"?

Keith
________________________________

**Comment by keithbriggs** [comment]
PS: why isn't math.prod better than reduce(mul,...)?

Keith
________________________________

**Comment by michael-123123** [comment]
Hi Keith - I will change the documentation as you suggested.
math.prod is only from python >= 3.8 (so not compatible with 3.7)
you're right about the else in line 168 -- will also remove.

**Comment by michael-123123** [comment]
In any case, now one test fails (test_out_of_bounds) because the output of size is expected to be numpy.int64, which is compared to (0,) instead of an int which is compared to 0.

i.e.,
 ```
exp = self.x[:-100]
self.assertEqual(exp.size, (0,))  # here is the problem
self.assertItemsAlmostEqual(exp.value, np.array([]))
```

The question is whether to fix the test - or wrap the value returned from size_from_shape with a numpy.int64?

**Comment by keithbriggs** [comment]
Then could the py version be tested and the best available function used? (Not on every call, obviously.)

**Comment by michael-123123** [comment]
@keithbriggs - yes, it could be tested (when the module loads).
However, I only have 3.7 to test and benchmark this fix.
I don't know whether math.prod is as fast as reduce, and I do know that reduce is as fast as any other solution we have.
so - not sure about it.

**Comment by keithbriggs** [comment]
Understood. I'm pretty sure math.prod will be faster when available, as it is less generic.

**Comment by michael-123123** [comment]
In that case - if I have time - I may set up an environment with python 3.8 to test it tomorrow.
Either way - I'm very happy with the 3x improvement caused by reduce.

**Comment by phschiele** [comment]
@michael-123123 Thanks for the PR!
Just for comparison, I have added `math.prod` to the chart. As @keithbriggs suggested, it is slightly faster. I would be in favor of not adding additional complexity dependent on the Python version for the small additional performance gain, though.

image

**Comment by michael-123123** [comment]
@phschiele Thanks.
I also think that the math.prod version is not fast enough to justify the added complexity.

**Review by SteveDiamond — APPROVED** [review]

**Pull request merged**